teps/0038-generic-workspaces.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Generic Workspaces
+authors:
+  - "@sbwsg"
+creation-date: 2020-12-11
+last-updated: 2020-12-11
+status: proposed
+---
+
+# TEP-0038: Generic Workspaces
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+<!-- /toc -->
+
+## Summary
+
+This doc describes a problem facing the Workspaces feature in Tekton Pipelines: there are many different
+ways that people and companies want to manage Workspaces but Tekton provides only a very narrow set of
+supported storage options.
+
+## Motivation
+
+Tekton Pipelines currently only offers very limited support for different types of volume in
+Workspaces: `emptyDir`, `configMap`, `secret`, `persistentVolumeClaim` and `volumeClaimTemplate`.
+
+The Pipelines API Spec is intentionally even more limited: only support for `emptyDir` is required
+to conform to the v1beta1 API. All other workspace types are optional. This is done to reduce
+implementation burden that would be imposed on proprietary platforms if they want to build
+Tekton-compliant integrations or services.
+
+For end-users this all means they are limited to a specific set of Tekton-supported volume types
+in their CI/CD runs. This negatively impacts many different types of user or potential user:
+- Users who want to use newer volume types added by Kubernetes like `projectedVolume`.
+- Orgs that mandate use of specific volume types Tekton doesn't support like `flexVolume`.
+- Orgs who write their own custom storage handling and manage everything themselves.
+
+For Platform Builders hoping to rely on the API Spec as a way to integrate with other
+Tekton-conforming systems they're let down by the Spec's only required workspace type
+being `emptyDir` while all others are marked optional. This means that:
+- A Platform Builder can't guarantee the types of volumes they'll receive when accepting
+TaskRuns from other "conforming" systems.
+- A Platform Builder has to handle the edge case that other conforming systems might not
+accept a TaskRun they submit with an optional volume type.
+- A Platform Builder can't limit the set of supported volume types to only those they want to allow.
+- A Platform Builder can't expand the set of supported volume types to add those they manage themselves.
+- A Platform Builder can expect the set of supported Workspace types to grow as support is added
+for new Kubernetes types (like `ProjectedVolume`) and must plan around those potential changes.
```

Comment:
i think im being pedantic so feel free to ignore me, but im wondering if 'platform developers' are the folks who have this set of concerns, or if that's more [tool developers](https://github.com/tektoncd/community/blob/main/user-profiles.md#5-supporting-tool-developer)? (i think the thing tripping me up is that im not sure why platform builders really care about systems other than the one they're building)

Reply:
> (i think the thing tripping me up is that im not sure why platform builders really care about systems other than the one they're building)

My take is they care in the sense that they want to interoperate with those systems via the Tekton API.  Resources from one exported and then imported to another.  If the Tekton API is that locus of integration then this is the list of responsibilities that falls on the platform developer's shoulders.

Does that interpretation make sense or do you still think these concerns are first and foremost those of Tool Developers?  Happy to change it if so.

Reply:
I've trimmed the list back a bit here in response to the feedback.

Reply:
kk sounds fine to me! maybe we could add tool developers as well? im thinking something like "as a tool developer I want to develop tools that will work with any k8s system" - for example the CLI can run Pipelines, but maybe it only knows about PVCs and emptyDirs as possible volumes, and doesn't know how to make a `Gluster` volume - id like my tool to be able to work with even systems that require  `Gluster` volumes

(i think im just trying to polish something that's already in great shape)

Reply:
Great point - makes total sense that this is also a concern of folks writing interoperable tools.  Updated!